### PR TITLE
New Restart-HCXManager Cmdlet

### PR DESCRIPTION
**Pull request by Asummers-MSFT**
Problem
Customers need to be able to restart hcx using run commands.
Solution
Create a cmdlet that restarts the hcx manager on a tenant's sddc.

Testing
MicrosoftTeams-image (11)
MicrosoftTeams-image (12)

Logs and metrics
n/a

Upcoming changes
n/a

### CODE REVIEW

File: Microsoft.AVS.Management/Microsoft.AVS.Management.psm1 [hunk 1]
```diff
@@ -1034,39 +1044,198 @@ function Set-ClusterDefaultStoragePolicy {
...
+    .Parameter HardReboot
+    Warning: This Parameter should be used as a last ditch effort where a soft-reboot wouldn't work.
+    Hard Reboots the VM instead of restarting the Guest OS.
+    .Parameter Timeout
+    Number of seconds the script is allowed to wait for sucessful connection to the hcx appliance before timing out.
+    .Example
+    # Skips Migrations and replications and hard reboots the system.
+    Restart-HcxManager -Force -HardReboot
+#>
+function Restart-HCXManager {
```
Comment (et1975) [inline Microsoft.AVS.Management/Microsoft.AVS.Management.psm1:1102]: Use `AVSAttribute`.

File: Microsoft.AVS.Management/HcxUtils.ps1 [hunk 2]
```diff
@@ -106,3 +106,25 @@ function Get-HcxMetaData {
...
+    $HcxVm = $null
+    $VmsList = Get-VM
+
+    foreach ($Vm in $VmsList) {
+        if($Vm.Name.Contains("HCX-MGR")) {
+            $HcxVm = $Vm
+            break
+        }
+    }
+    if(-not $HcxVm) {
```
Comment (et1975) [inline Microsoft.AVS.Management/HcxUtils.ps1:126]: I'd prefer that this check was done outside this function. For one thing throwing from here wouldn't stop the caller running, for another it's better to let the called decide what exactly they want to do about the absence of the VM - it makes this function easier to reason about and more useful.

File: Microsoft.AVS.Management/Microsoft.AVS.Management.psm1 [hunk 3]
```diff
@@ -1034,39 +1044,198 @@ function Set-ClusterDefaultStoragePolicy {
...
+        $privileges = @("VirtualMachine.Interact.PowerOff",
+            "VirtualMachine.Interact.PowerOn")
+        $HcxAdminCredential = New-TempUser -privileges $privileges -userName $UserName -userRole $UserRole
+
+        Write-Host "INPUTS: HardReboot=$HardReboot, Force=$Force, Port=$Port, Timeout=$Timeout"
+
+        $HcxServer = 'hcx'
+        $hcxVm = Get-HcxManagerVM
+
+        Add-UserToGroup -userName $UserName -group $Group
```
Comment (et1975) [inline Microsoft.AVS.Management/Microsoft.AVS.Management.psm1:1139]: I don't believe belonging to CloudAdmins will let this user access our management resources.

**Review by et1975 — APPROVED** [review]

File: Microsoft.AVS.Management/Microsoft.AVS.Management.psm1 [hunk 4]
```diff
@@ -1034,39 +1044,248 @@ function Set-ClusterDefaultStoragePolicy {
...
+                Restart-VM -VM $hcxVm -Confirm:$false | Out-Null
+                Write-Host "$($hcxVm.Name)'s powerstate=$($hcxVm.PowerState)"
+            }
+            else {
+                Write-Host "Restarting Guest OS..."
+                Restart-VMGuest -VM $hcxVm | Out-Null
+                Write-Host "$($hcxVm.Name)'s powerstate=$($hcxVm.PowerState)"
+            }
+        }
+        Write-Host "Waiting on connection to HCX appliance..."
```
Comment (jtrammell-msft) [inline Microsoft.AVS.Management/Microsoft.AVS.Management.psm1:1265]: Display how long its going to wait.

Comment (jtrammell-msft) [reply Microsoft.AVS.Management/Microsoft.AVS.Management.psm1:1265]: This is not a showstopper

**Pull request merged**